File: frontend/src/services/reminderService.ts
import { reminderRepository, syncService } from './repositories';
import type { Reminder } from '../types';

export const reminderService = {
  async getReminders(patientId: string, date?: string): Promise<Reminder[]> {
    return await reminderRepository.getReminders(patientId, date);
  },

  async getReminderById(id: string): Promise<Reminder | null> {
    return await reminderRepository.getReminderById(id);
  },

  async markAsTaken(id: string): Promise<Reminder> {
    const updated = await reminderRepository.updateReminderStatus(id, 'completed');
    await syncService.queueEvent({
      entityType: 'reminder',
      action: 'update',
      payload: updated,
    });
    return updated;
  },

  async snoozeReminder(id: string): Promise<Reminder> {
    const updated = await reminderRepository.updateReminderStatus(id, 'snoozed');
    await syncService.queueEvent({
      entityType: 'reminder',
      action: 'update',
      payload: updated,
    });
    return updated;
  },

  async createReminder(reminder: Omit<Reminder, 'id'>): Promise<Reminder> {
    const created = await reminderRepository.createReminder(reminder);
    await syncService.queueEvent({
      entityType: 'reminder',
      action: 'create',
      payload: created,
    });
    return created;
  },

  async updateReminder(id: string, data: Partial<Reminder>): Promise<Reminder> {
    const updated = await reminderRepository.updateReminder(id, data);
    await syncService.queueEvent({
      entityType: 'reminder',
      action: 'update',
      payload: updated,
    });
    return updated;
  },

  async deleteReminder(id: string): Promise<boolean> {
    const success = await reminderRepository.deleteReminder(id);
    await syncService.queueEvent({
      entityType: 'reminder',
      action: 'delete',
      payload: { id },
    });
    return success;
  }
};

File: backend/app/services/reminder_service.py
from typing import List, Optional
from datetime import date, datetime, timezone
from sqlalchemy.orm import Session
from app.models.reminder import Reminder, Medication
from app.repositories.reminder_repo import ReminderRepository, MedicationRepository
from app.schemas.reminder import ReminderCreate, ReminderUpdate
from app.core.exceptions import NotFoundException


class ReminderService:
    def __init__(self, db: Session):
        self.db = db
        self.reminder_repo = ReminderRepository(db)
        self.med_repo = MedicationRepository(db)

    def get_reminders(
        self, patient_id: str, scheduled_date: Optional[date] = None
    ) -> List[Reminder]:
        return self.reminder_repo.get_by_patient(patient_id, scheduled_date)

    def get_reminder_by_id(self, id: str) -> Reminder:
        rem = self.reminder_repo.get(id)
        if not rem:
            raise NotFoundException("Reminder")
        return rem

    def create_reminder(self, reminder_data: ReminderCreate) -> Reminder:
        return self.reminder_repo.create(reminder_data)

    def update_reminder(self, id: str, update_data: ReminderUpdate) -> Reminder:
        rem = self.get_reminder_by_id(id)
        return self.reminder_repo.update(rem, update_data)

    def update_status(self, id: str, status: str) -> Reminder:
        rem = self.get_reminder_by_id(id)
        rem.status = status
        if status == "completed":
            rem.completed_at = datetime.now(timezone.utc)
        self.db.commit()
        self.db.refresh(rem)
        return rem

    def delete_reminder(self, id: str) -> bool:
        return self.reminder_repo.delete(id)

    def get_medications(self, patient_id: str) -> List[Medication]:
        return self.med_repo.get_by_patient(patient_id)
